### PR TITLE
FOGL-484: Statistics Unit-Testing 

### PR DESCRIPTION
**Pull request by oshadmon**
The following is unit-testing for statistics.py and statistics_history.py

### CODE REVIEW

**Comment by oshadmon** [comment]
@ashish-ScaleDB - The one thing that might be missing is since I changed the name of statistics_history.py (from update_statistics_history.py), I should probably also update the files that call statistics_history.... The only issue is, I'm not 100% sure which files those are. 

Thanks!

**Comment by ashwinscale** [comment]
@vaibhav-ScaleDB @ashish-ScaleDB please review for unit test consistency, best practices, proper decorators(allure report)

**Comment by ashwinscale** [comment]
update this line to reflect change in statistics_history

https://github.com/foglamp/FogLAMP/blob/b28018c4ab9c774bcc5268764e1d6817db103136/src/sql/foglamp_init_data.sql#L119

File: src/python/tests/test_statistics_history.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+                    await conn.execute(stmt)
+        except Exception:
+            print('Updated failed: %s' % stmt)
+            raise
+        values[key] =+ 1
+    return values
+
+
+@pytest.mark.asyncio
+async def test_init_data():
```
Comment (ashish-ScaleDB) [inline src/python/tests/test_statistics_history.py]: Please add below annotations for all test cases in a file.
```
@pytest.allure.feature("unit")
@pytest.allure.story("statistics history")
```

Also, you may test and generate allure report on local itself after make changes. Please have a look on `docs->build-tools.rst`

File: src/python/tests/test_statistics_history.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+                async with engine.acquire() as conn:
+                    await conn.execute(stmt)
+        except Exception:
+            print('INSERT failed')
+            raise
+        values[key] = 0
+    return values
+
+
+async def _update_statistics_value(values: dict={None:None})->dict:
```
[the remark this answers is not in the document]
Comment (ashish-ScaleDB) [reply src/python/tests/test_statistics_history.py]: `values: dict={None:None}` is an Anti-pattern stuff, Default argument value is mutable. So best practice is to use a sentinel value to denote an empty list or dictionary.

**Comment by oshadmon** [comment]
@ashish-ScaleDB & @ashwinscale  - Both the pylint issues, and changes in `FogLAMP/src/sql/foglamp_init_data.sql` have been pushed into branch/PR.

**Comment by praveen-garg** [comment]
Please fix things with some conceptual references from https://github.com/foglamp/FogLAMP/pull/185#pullrequestreview-64245377

File: src/python/tests/test_statistics.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                _STATISTICS_TBL.c.key == key)
+            try:
+                async with aiopg.sa.create_engine(_CONNECTION_STRING) as engine:
+                    async with engine.acquire() as conn:
+                        async for result in conn.execute(stmt):
+                            assert result[0] == rand_value
+            except Exception:
+                raise
+
+            stmt = sa.select([_STATISTICS_TBL.c.previous_value]).select_from(_STATISTICS_TBL).where(
```
Comment (praveen-garg) [inline src/python/tests/test_statistics.py]: do we need a new call? can we fetch this c.previous_value in previous select itself? and compare?

File: src/python/tests/test_statistics.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        """
+        for key in _KEYS:
+            rand_value = random.randint(1, 10)
+            await update_statistics_value(statistics_key=key, value_increment=rand_value)
+            stmt = sa.select([_STATISTICS_TBL.c.value]).select_from(_STATISTICS_TBL).where(
+                _STATISTICS_TBL.c.key == key)
+            try:
+                async with aiopg.sa.create_engine(_CONNECTION_STRING) as engine:
+                    async with engine.acquire() as conn:
+                        async for result in conn.execute(stmt):
```
Comment (praveen-garg) [inline src/python/tests/test_statistics.py]: execute once? fetchRows and loop through that

File: src/python/tests/test_statistics.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                            assert result[0] == sum(rand_value)
+                            assert result[1] == 0
+            except Exception:
+                raise
+
+    async def test_invalid_statistics_key(self):
```
Comment (praveen-garg) [inline src/python/tests/test_statistics.py]: looks invalid test to me!

File: src/python/tests/test_statistics_history.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+pytestmark = pytest.mark.asyncio
+
+async def set_in_keys():
+    """Set statistics.keys column into a list to be used by test cases"""
+    stmt = sa.select([_STATS_TABLE.c.key])
+    try:
+        async with aiopg.sa.create_engine(_CONNECTION_STRING) as engine:
+            async with engine.acquire() as conn:
+                async for result in conn.execute(stmt):
+                    _KEYS.append(result[0].replace(" ", ""))
```
Comment (praveen-garg) [inline src/python/tests/test_statistics_history.py]: you can use strip() to TRIM

File: src/python/tests/test_statistics_history.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        for key in _KEYS:
+            stmt = sa.select([_STATS_TABLE.c.value, _STATS_TABLE.c.previous_value]).where(
+                _STATS_TABLE.c.key == key)
+            rand_value = random.randint(1, 10)
+            _update_previous_value(key=key, value=rand_value)
+            try:
+                async with aiopg.sa.create_engine(_CONNECTION_STRING) as engine:
+                    async with engine.acquire() as conn:
+                        async for result in conn.execute(stmt):
+                            assert result[0] == 0
```
Comment (praveen-garg) [inline src/python/tests/test_statistics_history.py]: it resets the value to 0 ?

File: src/python/tests/test_statistics_history.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    async def test_select_from_statistics(self):
+        """
+        Test that _select_from_statistics retrieves value and previous value from statistics table
+        :assert:
+            Assert that both value and previous value == 0
+        """
+        for key in _KEYS:
+            result = _select_from_statistics(key=key)
+            assert result[0][0] == 0
+            assert result[0][0] == result[0][1]
```
Comment (praveen-garg) [inline src/python/tests/test_statistics_history.py]: fine! result[0][1] == 0 itself

File: src/python/tests/test_statistics_history.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+    async def test_stats_history_main(self):
```
Comment (praveen-garg) [inline src/python/tests/test_statistics_history.py]: I'm not clear what this test is doing?! we can make it simpler with non random values and may be break into 2-3 tests?

File: src/sql/foglamp_init_data.sql [hunk 10]
```diff
@@ -117,7 +117,7 @@ INSERT INTO foglamp.statistics ( key, description, value, previous_value )
 
 insert into foglamp.scheduled_processes (name, script) values ('device', '["python3", "-m", "foglamp.device"]');
 insert into foglamp.scheduled_processes (name, script) values ('purge', '["python3", "-m", "foglamp.data_purge"]');
-insert into foglamp.scheduled_processes (name, script) values ('stats collector', '["python3", "-m", "foglamp.update_statistics_history"]');
+insert into foglamp.scheduled_processes (name, script) values ('stats collector', '["python3", "-m", "foglamp.statistics_history"]');
```
Comment (praveen-garg) [inline src/sql/foglamp_init_data.sql]: is the file name only being used here?

File: src/python/tests/test_statistics.py (hunk 5 above)
Comment (oshadmon) [reply src/python/tests/test_statistics.py]: @praveen-garg the test is invalid... That said, I'm wondering if an error should be returned when trying to update a statistics value (`update_statistcs_value`) with `statistics_key` that does not exist? Right now when doing such a thing, then it's as if nothing happened. 
CC @ashwinscale @MarkRiddoch

File: src/python/tests/test_statistics_history.py (hunk 7 above)
Comment (oshadmon) [reply src/python/tests/test_statistics_history.py]: By calling `_update_previous_value`, the code just updates the `previous_value` for a given key. It does not update `value`. As such `value` == 0. 
When relooking at the code, I agree that doing `assert result[0] == 0` can be skipped

File: src/sql/foglamp_init_data.sql (hunk 10 above)
Comment (oshadmon) [reply src/sql/foglamp_init_data.sql]: From my understanding yes.

**Comment by oshadmon** [comment]
@praveen-garg & @vaibhav-ScaleDB - Implemented the following changes
* try/except removed.
* Simplified tests by doing a check on a single key rather than by iterating through all of them. This is because there isn't a behavioral difference based on the key.

**Comment by vaibhav-ScaleDB** [comment]
Please rebase and resolve conflicts.

**Comment by izoratti-dianomic** [comment]
There should not be files changing the docs at the moment, I am working at the documentation for the product and we did not consider to add this info in it (it will be added, eventually). 
Simply remove the conflicting files for the moment (or avoid autogeneration if this is the case now).

File: src/python/tests/test_statistics_history.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        async with aiopg.sa.create_engine(_CONNECTION_STRING) as engine:
+            async with engine.acquire() as conn:
+                # Execute select_stats_stmt
+                async for result in conn.execute(select_stats_stmt):
+                    select_stats = [result[0], result[1]]
+                # Execute select_stats_history_stmt
+                async for result in conn.execute(select_stats_history_stmt):
+                    select_stats_history = result[0]
+
+        assert select_stats[0] == select_stats[1] == select_stats_history
```
Comment (vaibhav-ScaleDB) [inline src/python/tests/test_statistics_history.py]: pycharm error: Local variable 'select_stats' and 'select_stats_history' might be referenced before assignment.

**Comment by oshadmon** [comment]
@praveen-garg - Please rereview. 
Thanks!

**Comment by praveen-garg** [comment]
squash these, please!

**Comment by praveen-garg** [comment]
Better to close this and open 2 new PR:
1) Tests for statistics_history https://github.com/foglamp/FogLAMP/pull/252 *remove __main__ tests
2) Tests for statistics

**Comment by MarkRiddoch** [comment]
This PR has a long history and I suspect the base code has moved on somewhat. We should erase this with the latest code and try to complete this review in the net few days so that it doesn't miss the code restructuring.

**Comment by oshadmon** [comment]
Closed the PR, and will reopen it, as two separate PRs 1 for statistics and another for statistics_history.

**Pull request closed without merging**